File: day2.js
fs = require('fs');

function day2(input) {
    let sum = 0

    for(line of input) {
        let min = Number.MAX_SAFE_INTEGER
        let max = -1
        for(n of line) {
            min = Math.min(n,min)
            max = Math.max(n,max)
        }
        sum += max - min
    }
    return sum
}


function day2_part2(input) {
    let sum = 0
    for(line of input) {
        line.some( (value, index) => {
            for (let i = index + 1; i<line.length; i++) {

                if(value > line[i]) { v1 = value; v2 = line[i] } 
                else { v1 = line[i]; v2 = value }

                if( v1%v2 == 0 ){
                    sum += v1 / v2
                    return true
                }
            }
        })
    }
    return sum
}


fs.readFile('input/day2','utf-8', (err, data) => { 
    const lines = data.split("\n")
    let array = []
    for(l of lines) {
        const values = l.split(" ").filter(i => i)
        let line = []
        values.map(v => { line.push(Number(v)) })
        array.push(line)
    }
    
    console.log(day2(array))
    console.log(day2_part2(array))
})

File: aoc_2023/day2.py
from pathlib import Path
import math

def part1(data, cubes):
    def is_valid(game):
        return all(a <= b for round in game for a,b in zip(round, cubes))
    return sum([id for id, game in enumerate(data, 1) if is_valid(game)])


def part2(data):
    def get_power(game):
        return math.prod([max(i) for i in zip(*game)])
    return sum([get_power(game) for game in data])


def parse(file):
    '''returns [  
        [[r,g,b],[r,g,b]...] ,
        ...   
    ]'''
    def rgb(str):
        result = [0,0,0]
        for cubes in str.split(', '):
            v, c = cubes.split(' ')
            result[{'red':0, 'green':1, 'blue':2}[c]] = int(v)
        return result

    return [
        [rgb(x) for x in line.split(': ')[1].split('; ')]
        for line in file.split('\n')
    ]


if __name__ == "__main__":
    with Path(__file__).parent.joinpath("input/"+Path(__file__).stem).open(encoding="UTF-8") as f:
        DATA = parse(f.read())

        print(f"Phase 1: {part1(DATA, [12,13,14])}")
        print(f"Phase 2: {part2(DATA)}")
